scripts/ci-docker-images/builder-x86_64-unknown-linux-musl/Dockerfile

```diff
@@ -1,19 +1,343 @@
+# and
+# https://renenyffenegger.ch/notes/development/languages/C-C-plus-plus/GCC/options/no/compare-nostartfiles-nodefaultlibs-nolibc-nostdlib.
+ENV MUSL_CFLAGS="${LTO:+-f$LTO} -nostdinc -isystem $LIBS_PREFIX/include"
+ENV MUSL_CXXFLAGS="$MUSL_CFLAGS -nostdinc++ -I$LIBS_PREFIX/include/c++/v1"
+
+ENV MUSL_LDFLAGS="-fuse-ld=lld -static -nostdlib -nostartfiles -L$LIBS_PREFIX/lib -L$LIBS_PREFIX/lib/linux -lc++ -lc++abi -lunwind -lclang_rt.builtins-x86_64 -lc"
+ENV MUSL_STARTFILES="$LIBS_PREFIX/lib/crt1.o $LIBS_PREFIX/lib/linux/clang_rt.crtbegin-x86_64.o $LIBS_PREFIX/lib/linux/clang_rt.crtend-x86_64.o"
+
+RUN mkdir -p $LIBS_PREFIX/bin
+RUN echo \
+"#!/bin/sh\n"\
+"case \"\$@\" in *-shared*);; *-nostdlib*);; *) STARTFILES=\"$MUSL_STARTFILES\";; esac\n"\
+"$CLANG_PREFIX/bin/clang -Qunused-arguments $MUSL_CFLAGS \$@ \$STARTFILES $MUSL_LDFLAGS\n"\
+"exit \$?" > $LIBS_PREFIX/bin/musl-cc
+RUN echo \
+"#!/bin/sh\n"\
+"case \"\$@\" in *-shared*);; *-nostdlib*);; *) STARTFILES=\"$MUSL_STARTFILES\";; esac\n"\
+"$CLANG_PREFIX/bin/clang++ -Qunused-arguments $MUSL_CXXFLAGS \$@ \$STARTFILES $MUSL_LDFLAGS\n"\
+"exit \$?" > $LIBS_PREFIX/bin/musl-c++
+RUN echo \
+"#!/bin/sh\n"\
+"$CLANG_PREFIX/bin/ld.lld -L$LIBS_PREFIX/lib -L$LIBS_PREFIX/lib/linux \$@\n"\
+"exit \$?" > $LIBS_PREFIX/bin/musl-ld
+RUN chmod +x $LIBS_PREFIX/bin/*
+
+# At this point a fully functional C++ compiler that is able to produce
+# static binaries linked with musl and libc++ is bootstrapped.
+# It can be used by calling musl-c++ (or musl-cc for C) executable.
+# However, we need to also create generic aliases to make it possible to
+# use it as a drop-in replacement for the system-wide GCC.
+RUN ln -s $LIBS_PREFIX/bin/musl-cc $LIBS_PREFIX/bin/cc
+RUN ln -s $LIBS_PREFIX/bin/musl-cc $LIBS_PREFIX/bin/gcc
+RUN ln -s $LIBS_PREFIX/bin/musl-cc $LIBS_PREFIX/bin/musl-gcc
+RUN ln -s $LIBS_PREFIX/bin/musl-c++ $LIBS_PREFIX/bin/c++
+RUN ln -s $LIBS_PREFIX/bin/musl-c++ $LIBS_PREFIX/bin/g++
+RUN ln -s $LIBS_PREFIX/bin/musl-c++ $LIBS_PREFIX/bin/musl-g++
+RUN ln -s $LIBS_PREFIX/bin/musl-ld $LIBS_PREFIX/bin/ld
+
+# Some build systems hardcode -lstdc++ linker flag on all Linux systems.
+# Because our linker is already configured to link with libc++ instead,
+# we can just provide dummy libstdc++ to be compatible with GNU systems.
+# For example, macOS provides similar experience, where
+# `clang++ -o program program.cpp -lstdc++` would still link to libc++.
+RUN echo > dummy.c && \
+  $CLANG_PREFIX/bin/clang ${LTO:+-f$LTO} -c dummy.c && \
+  $CLANG_PREFIX/bin/ar cr $LIBS_PREFIX/lib/libstdc++.a dummy.o && \
+  rm dummy.c dummy.o
+
+# Install LibreSSL.
+WORKDIR $LIBRESSL_DIR
+# Rust crate openssl-sys supports only released versions of LibreSSL.
+# In order to be able to use the latest stable release with musl,
+# we need to backport unreleased changes from this PR:
+# https://github.com/libressl-portable/portable/pull/529/files.
```

Comment:
what are the implications of this? Is there any follow up that we might need to do here?

Reply:
`program_invocation_short_name` is a global variable that contains short name of the program. It is not a part of standard C library, but a GNU extension. It is implemented by `musl` though. The original code in LibreSSL checks on Linux whether C library is `glibc` and generates a compile-time error if it is not, even if `program_invocation_short_name` is available. With the patch the error is not thrown and `program_invocation_short_name` from `musl` is used.

Because the issue is fixed in upstream by the mentioned PR, it should be possible to remove this patching after the next release of LibreSSL.